crates/ra_ide/src/hover.rs

```diff
@@ -320,6 +332,270 @@ fn hover_text_from_name_kind(db: &RootDatabase, def: Definition) -> Option<Strin
+        // FIXME:
+        Some(Namespace::Macros) => None?,
+    };
+
+    // Get the filepath of the final symbol
+    let def: ModuleDef = defid.into();
+    let module = def.module(db)?;
+    let krate = module.krate();
+    let ns = match namespace {
+        Namespace::Types => ItemInNs::Types(defid),
+        Namespace::Values => ItemInNs::Values(defid),
+        // FIXME:
+        Namespace::Macros => None?,
+    };
+    let import_map = db.import_map(krate.into());
+    let path = import_map.path_of(ns)?;
+
+    Some(
+        get_doc_url(db, &krate)?
+            .join(&format!("{}/", krate.display_name(db)?))
+            .ok()?
+            .join(&path.segments.iter().map(|name| format!("{}", name)).join("/"))
+            .ok()?
+            .join(&get_symbol_filename(db, &Definition::ModuleDef(def))?)
+            .ok()?
+            .into_string(),
+    )
+}
+
+/// Try to resolve path to local documentation via path-based links (i.e. `../gateway/struct.Shard.html`).
+fn try_resolve_path(db: &RootDatabase, definition: &Definition, link: &str) -> Option<String> {
+    if !link.contains("#") && !link.contains(".html") {
+        return None;
+    }
+    let ns = if let Definition::ModuleDef(moddef) = definition {
+        ItemInNs::Types(moddef.clone().into())
+    } else {
+        return None;
+    };
+    let module = definition.module(db)?;
+    let krate = module.krate();
+    let import_map = db.import_map(krate.into());
+    let base = once(format!("{}", krate.display_name(db)?))
+        .chain(import_map.path_of(ns)?.segments.iter().map(|name| format!("{}", name)))
+        .join("/");
+
+    get_doc_url(db, &krate)
+        .and_then(|url| url.join(&base).ok())
+        .and_then(|url| {
+            get_symbol_filename(db, definition).as_deref().map(|f| url.join(f).ok()).flatten()
+        })
+        .and_then(|url| url.join(link).ok())
+        .map(|url| url.into_string())
+}
+
+/// Try to get the root URL of the documentation of a crate.
+// FIXME: Special case standard, core, alloc libraries
+fn get_doc_url(db: &RootDatabase, krate: &Crate) -> Option<Url> {
+    // Look for #![doc(html_root_url = "...")]
+    let attrs = db.attrs(AttrDef::from(krate.root_module(db)?).into());
+    let doc_attr_q = attrs.by_key("doc");
+    let doc_url = if doc_attr_q.exists() {
+        doc_attr_q.tt_values().filter_map(|tt| match tt.token_trees.as_slice() {
+            &[
+                TokenTree::Leaf(Leaf::Ident(Ident{text: ref ident_text, ..})),
+                TokenTree::Leaf(Leaf::Punct(Punct{r#char: '=', ..})),
+                TokenTree::Leaf(Leaf::Literal(Literal{ref text, ..}))
+            ] if ident_text == "html_root_url" => Some(text.to_string()),
+            _ => {
+                None
+            }
+        }).next()
+    } else {
+        // Fallback to docs.rs
+        // FIXME: Specify an exact version here (from Cargo.lock)
+        Some(format!("https://docs.rs/{}/*", krate.display_name(db)?))
```

Comment:
Is there an existing API to get the version of a `Crate`?

Reply:
[`cargo-lock`](https://docs.rs/cargo-lock/4.0.1/cargo_lock/index.html) looks like it'd do this. I'm not sure of the best way to add this though. Would it make sense to add this information to the world and pass it down to here?

Reply:
I've updated the note to mention the multiple versions of a crate possibility. Ultimately I think this can be left for now.